File: lib/src/main/java/otang/pulse/lib/VisualizerStreamHandler.java
package otang.pulse.lib;

import android.content.Context;
import android.media.audiofx.Visualizer;
import android.os.Handler;
import android.os.Looper;
import android.os.Message;
import android.util.Log;

public class VisualizerStreamHandler {

    public interface Listener {
        void onStreamAnalyzed(boolean isValid);

        void onFFTUpdate(byte[] bytes);

        void onWaveFormUpdate(byte[] bytes);
    }

    protected static final String TAG = VisualizerStreamHandler.class.getSimpleName();
    protected static final boolean ENABLE_WAVEFORM = false;
    protected static final int MSG_STREAM_VALID = 55;
    protected static final int MSG_STREAM_INVALID = 56;
    // we have 6 seconds to get three consecutive valid frames
    protected static final int VALIDATION_TIME_MILLIS = 6000;
    protected static final int VALID_BYTES_THRESHOLD = 3;
    protected Visualizer mVisualizer;
    // manage stream validation
    protected int mConsecutiveFrames;
    protected boolean mIsValidated;
    protected boolean mIsAnalyzed;
    protected boolean mIsPrepared;
    protected boolean mIsPaused;
    protected Context mContext;
    protected Listener mListener;

    private final Handler mHandler = new Handler(Looper.getMainLooper()) {
        @Override
        public void handleMessage(Message m) {
            switch (m.what) {
                case MSG_STREAM_VALID:
                    mIsAnalyzed = true;
                    mIsValidated = true;
                    mIsPrepared = false;
                    mListener.onStreamAnalyzed(true);
                    break;
                case MSG_STREAM_INVALID:
                    mIsAnalyzed = true;
                    mIsValidated = false;
                    mIsPrepared = false;
                    mListener.onStreamAnalyzed(false);
                    break;
            }
        }
    };

    public VisualizerStreamHandler(Context context, VisualizerStreamHandler.Listener listener) {
        mContext = context;
        mListener = listener;
    }

    /**
     * Links the visualizer to a player
     */
    public final void link(int id) {
        pause();
        resetAnalyzer();
        mVisualizer = null;
        try {
            mVisualizer = new Visualizer(id);
        } catch (Exception e) {
            Log.e(TAG, "Error enabling visualizer!", e);
            return;
        }
        mVisualizer.setEnabled(false);
        mVisualizer.setCaptureSize(Visualizer.getCaptureSizeRange()[1]);
        Visualizer.OnDataCaptureListener captureListener = new Visualizer.OnDataCaptureListener() {
            @Override
            public void onWaveFormDataCapture(Visualizer visualizer, byte[] bytes, int samplingRate) {
                if (ENABLE_WAVEFORM) {
                    analyze(bytes);
                    if (isValidStream() && !mIsPaused) {
                        mListener.onWaveFormUpdate(bytes);
                    }
                }
            }

            @Override
            public void onFftDataCapture(Visualizer visualizer, byte[] bytes, int samplingRate) {
                analyze(bytes);
                if (isValidStream() && !mIsPaused) {
                    mListener.onFFTUpdate(bytes);
                }
            }
        };
        mVisualizer.setDataCaptureListener(captureListener, (int) (Visualizer.getMaxCaptureRate() * 0.75), ENABLE_WAVEFORM, true);
        mVisualizer.setEnabled(true);
    }

    public final void unlink() {
        if (mVisualizer != null) {
            pause();
            mVisualizer.setEnabled(false);
            mVisualizer.release();
            mVisualizer = null;
            resetAnalyzer();
        }
    }

    public boolean isValidStream() {
        return mIsAnalyzed && mIsValidated;
    }

    public void resetAnalyzer() {
        mIsAnalyzed = false;
        mIsValidated = false;
        mIsPrepared = false;
        mConsecutiveFrames = 0;
    }

    public void pause() {
        mIsPaused = true;
    }

    public void resume() {
        mIsPaused = false;
    }

    private void analyze(byte[] data) {
        if (mIsAnalyzed) {
            return;
        }
        if (!mIsPrepared) {
            mHandler.sendMessageDelayed(mHandler.obtainMessage(MSG_STREAM_INVALID), VALIDATION_TIME_MILLIS);
            mIsPrepared = true;
        }
        if (isDataEmpty(data)) {
            mConsecutiveFrames = 0;
        } else {
            mConsecutiveFrames++;
        }
        if (mConsecutiveFrames == VALID_BYTES_THRESHOLD) {
            mIsPaused = true;
            mHandler.removeMessages(MSG_STREAM_INVALID);
            mHandler.sendEmptyMessage(MSG_STREAM_VALID);
        }
    }

    private boolean isDataEmpty(byte[] data) {
        for (byte datum : data) {
            if (datum != 0) {
                return false;
            }
        }
        return true;
    }
}

File: lib/src/main/java/otang/pulse/lib/VisualizerStreamHandler.kt
package otang.pulse.lib

import android.media.audiofx.Visualizer
import android.media.audiofx.Visualizer.OnDataCaptureListener
import android.os.Handler
import android.os.Looper
import android.os.Message
import android.util.Log

class VisualizerStreamHandler(private var mListener: Listener) {
    interface Listener {
        fun onStreamAnalyzed(isValid: Boolean)
        fun onFFTUpdate(bytes: ByteArray?)
        fun onWaveFormUpdate(bytes: ByteArray?)
    }

    private var mVisualizer: Visualizer? = null

    // manage stream validation
    private var mConsecutiveFrames = 0
    private var mIsValidated = false
    private var mIsAnalyzed = false
    private var mIsPrepared = false
    private var mIsPaused = false
    private val mHandler: Handler = object : Handler(Looper.getMainLooper()) {
        override fun handleMessage(m: Message) {
            when (m.what) {
                MSG_STREAM_VALID -> {
                    mIsAnalyzed = true
                    mIsValidated = true
                    mIsPrepared = false
                    mListener.onStreamAnalyzed(true)
                }

                MSG_STREAM_INVALID -> {
                    mIsAnalyzed = true
                    mIsValidated = false
                    mIsPrepared = false
                    mListener.onStreamAnalyzed(false)
                }
            }
        }
    }

    /**
     * Links the visualizer to a player
     */
    fun link(id: Int) {
        pause()
        resetAnalyzer()
        mVisualizer = null
        mVisualizer = try {
            Visualizer(id)
        } catch (e: Exception) {
            Log.e(TAG, "Error enabling visualizer!", e)
            return
        }
        mVisualizer!!.enabled = false
        mVisualizer!!.captureSize = Visualizer.getCaptureSizeRange()[1]
        val captureListener: OnDataCaptureListener = object : OnDataCaptureListener {
            override fun onWaveFormDataCapture(
                visualizer: Visualizer,
                bytes: ByteArray,
                samplingRate: Int
            ) {
                if (ENABLE_WAVEFORM) {
                    analyze(bytes)
                    if (isValidStream && !mIsPaused) {
                        mListener.onWaveFormUpdate(bytes)
                    }
                }
            }

            override fun onFftDataCapture(
                visualizer: Visualizer,
                bytes: ByteArray,
                samplingRate: Int
            ) {
                analyze(bytes)
                if (isValidStream && !mIsPaused) {
                    mListener.onFFTUpdate(bytes)
                }
            }
        }
        mVisualizer!!.setDataCaptureListener(
            captureListener,
            (Visualizer.getMaxCaptureRate() * 0.75).toInt(),
            ENABLE_WAVEFORM,
            true
        )
        mVisualizer!!.enabled = true
    }

    fun unlink() {
        if (mVisualizer != null) {
            pause()
            mVisualizer!!.enabled = false
            mVisualizer!!.release()
            mVisualizer = null
            resetAnalyzer()
        }
    }

    val isValidStream: Boolean
        get() = mIsAnalyzed && mIsValidated

    private fun resetAnalyzer() {
        mIsAnalyzed = false
        mIsValidated = false
        mIsPrepared = false
        mConsecutiveFrames = 0
    }

    fun pause() {
        mIsPaused = true
    }

    fun resume() {
        mIsPaused = false
    }

    private fun analyze(data: ByteArray) {
        if (mIsAnalyzed) {
            return
        }
        if (!mIsPrepared) {
            mHandler.sendMessageDelayed(
                mHandler.obtainMessage(MSG_STREAM_INVALID),
                VALIDATION_TIME_MILLIS.toLong()
            )
            mIsPrepared = true
        }
        if (isDataEmpty(data)) {
            mConsecutiveFrames = 0
        } else {
            mConsecutiveFrames++
        }
        if (mConsecutiveFrames == VALID_BYTES_THRESHOLD) {
            mIsPaused = true
            mHandler.removeMessages(MSG_STREAM_INVALID)
            mHandler.sendEmptyMessage(MSG_STREAM_VALID)
        }
    }

    private fun isDataEmpty(data: ByteArray): Boolean {
        for (datum in data) {
            if (datum.toInt() != 0) {
                return false
            }
        }
        return true
    }

    companion object {
        private val TAG = VisualizerStreamHandler::class.java.simpleName
        private const val ENABLE_WAVEFORM = false
        private const val MSG_STREAM_VALID = 55
        private const val MSG_STREAM_INVALID = 56

        // we have 6 seconds to get three consecutive valid frames
        private const val VALIDATION_TIME_MILLIS = 6000
        private const val VALID_BYTES_THRESHOLD = 3
    }
}